src/Npgsql/NpgsqlConnector.cs

```diff
@@ -1042,14 +1042,22 @@ internal ValueTask<IBackendMessage> ReadMessage(bool async, DataRowLoadingMode d
                 }
 
                 PostgresException? error = null;
+                CancellationTokenSource? cts = null;
+                CancellationToken ct = default;
 
                 try
                 {
                     ReceiveTimeout = UserTimeout;
 
+                    if (async && UserTimeout > 0)
```

Comment:
We shouldn't set ReceiveTimeout (just above) when async is true, we can bring it into this condition.